hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java

```diff
@@ -16,58 +16,190 @@
- *
- * @param <T> the type of {@link #payload}
+ * A event for discovery scans.
  *
  * @see InventoryListener
  */
 public class InventoryEvent<L> {
 
     private final SamplingService<L> samplingService;
-    private final List<Resource<L>> payload;
+    private final ResourceManager<L> resourceManager;
+    private final Optional<ResourceTypeManager<L>> resourceTypeManager;
+    private final Map<String, Resource<L>> addedOrModifiedRootResources;
+    private final Map<String, Resource<L>> removedRootResources;
+    private final List<Resource<L>> addedOrModified;
+    private final List<Resource<L>> removed;
 
     /**
      * Creates an inventory event.
-     *
      * @param samplingService a service that provides details such as feed ID and endpoint information that helps
      *                        identify the resources in the event, plus has methods that can be used to monitor
      *                        the resources in the event.
-     * @param payload the list of resources associated with this event
+     * @param resourceManager the resources associated with the event
+     * @param resourceTypeManager the resource types associated with the event. Omit if resource type sync is not
+     *                           needed
+     * @param addedOrModified list of added or modified resources
+     * @param removed         list of removed resources
      */
-    public InventoryEvent(SamplingService<L> samplingService, List<Resource<L>> payload) {
+    private InventoryEvent(SamplingService<L> samplingService,
+                          ResourceManager<L> resourceManager,
+                          Optional<ResourceTypeManager<L>> resourceTypeManager,
+                          List<Resource<L>> addedOrModified,
+                          List<Resource<L>> removed) {
         if (samplingService == null) {
             throw new IllegalArgumentException("Sampling service cannot be null");
         }
 
-        if (payload == null) {
-            payload = Collections.emptyList();
+        if (resourceManager == null) {
+            throw new IllegalArgumentException("Resource manager cannot be null");
+        }
+
+        if (resourceTypeManager == null) {
+            throw new IllegalArgumentException("Resource type manager cannot be null");
         }
 
         this.samplingService = samplingService;
-        this.payload = payload;
+        this.resourceManager = resourceManager;
+        this.resourceTypeManager = resourceTypeManager;
+        this.addedOrModified = addedOrModified;
+        this.removed = removed;
+
+        // Distribute 'addedOrModified' and 'removed' in 'addedOrModifiedRootResources' and 'removedRootResources'
+        addedOrModifiedRootResources = new HashMap<>();
+        removedRootResources = new HashMap<>();
+        addedOrModified.forEach(r -> {
+            Resource<L> root = getRootResource(r);
+            addedOrModifiedRootResources.put(root.getID().getIDString(), root);
```

Comment:
Yes, this is what is intended. The reason is that the whole root resource subtree is going to be persisted. So we need to separate root resources that have 1 modification (at least) wherever in their subtree = `addedOrModifiedRootResources` map ; and root resources that need to be entirely deleted = `removedRootResources` map.